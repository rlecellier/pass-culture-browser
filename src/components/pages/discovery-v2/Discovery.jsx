import PropTypes from 'prop-types'
import React, { Fragment, PureComponent } from 'react'
import { Route } from 'react-router-dom'

import DeckContainer from './Deck/DeckContainer'
import BookingContainer from '../../layout/Booking/BookingContainer'
import BookingCancellationContainer from '../../layout/BookingCancellation/BookingCancellationContainer'
import AbsoluteFooterContainer from '../../layout/AbsoluteFooter/AbsoluteFooterContainer'
import LoaderContainer from '../../layout/Loader/LoaderContainer'
import isDetailsView from '../../../utils/isDetailsView'
import isCancelView from '../../../utils/isCancelView'
import { MINIMUM_DELAY_BEFORE_UPDATING_SEED_3_HOURS } from './utils/utils'

class Discovery extends PureComponent {
  constructor(props) {
    super(props)

    this.state = {
      atWorldsEnd: false,
      hasError: false,
      hasError500: false,
      hasNoMoreRecommendations: false,
      isLoading: false,
    }
  }

  componentDidMount() {
    const {
      recommendations,
      redirectToFirstRecommendationIfNeeded,
      saveLastRecommendationsRequestTimestamp,
      shouldReloadRecommendations,
    } = this.props

    if (shouldReloadRecommendations) {
      this.updateRecommendations()
      saveLastRecommendationsRequestTimestamp()
    } else {
      redirectToFirstRecommendationIfNeeded(recommendations)
    }
  }

  componentDidUpdate(prevProps) {
    const {
      location,
      recommendations,
      redirectToFirstRecommendationIfNeeded,
      seedLastRequestTimestamp,
      updateLastRequestTimestamp,
    } = this.props
    const { location: prevLocation } = prevProps

    if (prevLocation.pathname !== location.pathname && location.pathname === '/decouverte') {
      redirectToFirstRecommendationIfNeeded(recommendations)
    }

    if (Date.now() > seedLastRequestTimestamp + MINIMUM_DELAY_BEFORE_UPDATING_SEED_3_HOURS) {
      updateLastRequestTimestamp()
    }
  }

  componentWillUnmount() {
    const { deleteTutorials, tutorials, resetRecommendations } = this.props

    if (tutorials.length > 0) {
      deleteTutorials(tutorials)
    }

    resetRecommendations()
  }

  handleFail = (state, action) => {
    this.setState({
      hasError: true,
      hasError500: action.payload.status === 500 ? true : false,
      hasNoMoreRecommendations: true,
      isLoading: false,
    })
  }

  handleSuccess = (state, action) => {
    const {
      recommendations,
      resetReadRecommendations,
      resetSeenOffers,
      redirectToFirstRecommendationIfNeeded,
    } = this.props

    const { data: loadedRecommendations = [] } = action && action.payload
    const atWorldsEnd = loadedRecommendations.length === 0
    const hasNoMoreRecommendations = (!recommendations || !recommendations.length) && atWorldsEnd

    this.setState({ atWorldsEnd, hasNoMoreRecommendations, isLoading: false }, () => {
      resetReadRecommendations()
      resetSeenOffers()
      redirectToFirstRecommendationIfNeeded(loadedRecommendations)
    })
  }

  updateRecommendations = () => {
    const {
      currentRecommendation,
      loadRecommendations,
      readRecommendations,
      seenOffers,
      recommendations,
      shouldReloadRecommendations,
    } = this.props

    const { atWorldsEnd, isLoading } = this.state
    if (atWorldsEnd || isLoading) {
      return
    }

    this.setState({ isLoading: true }, () => {
      loadRecommendations(
        this.handleSuccess,
        this.handleFail,
        currentRecommendation,
        recommendations,
        readRecommendations,
        seenOffers,
        shouldReloadRecommendations
      )
    })
  }

  renderBooking = () => {
    const { currentRecommendation } = this.props
    return <BookingContainer recommendation={currentRecommendation} />
  }

  renderBookingCancellation = () => <BookingCancellationContainer />

  renderDeck = () => <DeckContainer handleRequestPutRecommendations={this.updateRecommendations} />

  render() {
    const { match } = this.props
    const { hasError, hasError500, hasNoMoreRecommendations, isLoading } = this.state
    const cancelView = isCancelView(match)

    return (
      <Fragment>
        {!hasNoMoreRecommendations && (
          <main className="discovery-page no-padding page with-footer">
            <Route
              key="route-discovery-deck"
              path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details|transition)?/:booking(reservation)?/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?"
              render={this.renderDeck}
              sensitive
            />
            <Route
              key="route-discovery-booking"
              path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details)/:booking(reservation)/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?/:confirmation(confirmation)?"
              render={cancelView ? this.renderBookingCancellation : this.renderBooking}
              sensitive
            />
            <AbsoluteFooterContainer
              areDetailsVisible={isDetailsView(match)}
              borderTop
              id="deck-footer"
            />
          </main>
        )}
        <LoaderContainer
          hasError={hasError}
          hasError500={hasError500}
          hasNoMoreRecommendations={hasNoMoreRecommendations}
          isLoading={isLoading}
        />
      </Fragment>
    )
  }
}

Discovery.defaultProps = {
  currentRecommendation: null,
  readRecommendations: null,
  recommendations: null,
  seenOffers: null,
}

Discovery.propTypes = {
  currentRecommendation: PropTypes.shape(),
  deleteTutorials: PropTypes.func.isRequired,
  loadRecommendations: PropTypes.func.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      view: PropTypes.string,
    }),
  }).isRequired,
  readRecommendations: PropTypes.arrayOf(PropTypes.shape()),
  recommendations: PropTypes.arrayOf(PropTypes.shape()),
  redirectToFirstRecommendationIfNeeded: PropTypes.func.isRequired,
  resetReadRecommendations: PropTypes.func.isRequired,
  resetRecommendations: PropTypes.func.isRequired,
  resetSeenOffers: PropTypes.func.isRequired,
  saveLastRecommendationsRequestTimestamp: PropTypes.func.isRequired,
  seedLastRequestTimestamp: PropTypes.number.isRequired,
  seenOffers: PropTypes.arrayOf(PropTypes.shape()),
  shouldReloadRecommendations: PropTypes.bool.isRequired,
  tutorials: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  updateLastRequestTimestamp: PropTypes.func.isRequired,
}

export default Discovery
