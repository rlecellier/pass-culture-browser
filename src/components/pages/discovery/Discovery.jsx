import PropTypes from 'prop-types'
import React, { Fragment, PureComponent } from 'react'
import { Route, Switch } from 'react-router-dom'

import BookingContainer from '../../layout/Booking/BookingContainer'
import BookingCancellationContainer from '../../layout/BookingCancellation/BookingCancellationContainer'
import { ApiError } from '../../layout/ErrorBoundaries/ErrorsPage/ApiError'
import PageNotFoundContainer from '../../layout/ErrorBoundaries/ErrorsPage/PageNotFound/PageNotFoundContainer'
import LoaderContainer from '../../layout/Loader/LoaderContainer'
import DeckContainer from './Deck/DeckContainer'
import { MINIMUM_DELAY_BEFORE_UPDATING_SEED_3_HOURS } from './utils/utils'

class Discovery extends PureComponent {
  constructor(props) {
    super(props)

    this.state = {
      atWorldsEnd: false,
      isLoading: false,
    }
  }

  componentDidMount() {
    const {
      recommendations,
      redirectToFirstRecommendationIfNeeded,
      saveLastRecommendationsRequestTimestamp,
      shouldReloadRecommendations,
      coordinates,
      trackGeolocation,
    } = this.props

    const isGeolocated = coordinates && coordinates.latitude && coordinates.longitude
    if (isGeolocated) {
      trackGeolocation()
    }

    if (shouldReloadRecommendations) {
      this.updateRecommendations()
      saveLastRecommendationsRequestTimestamp()
    } else {
      redirectToFirstRecommendationIfNeeded(recommendations)
    }
  }

  componentDidUpdate(prevProps) {
    const {
      history,
      location,
      recommendations,
      redirectToFirstRecommendationIfNeeded,
      match,
      seedLastRequestTimestamp,
      updateLastRequestTimestamp,
    } = this.props
    const { location: prevLocation, recommendations: prevRecommendations } = prevProps

    if (prevLocation.pathname !== location.pathname && location.pathname === '/decouverte') {
      redirectToFirstRecommendationIfNeeded(recommendations)
    }

    if (Date.now() > seedLastRequestTimestamp + MINIMUM_DELAY_BEFORE_UPDATING_SEED_3_HOURS) {
      updateLastRequestTimestamp()
    }

    if (prevRecommendations !== recommendations) {
      const { params } = match
      const { offerId } = params
      const isOfferInReco = recommendations.filter(reco => reco.offerId === offerId).length > 0
      if (recommendations.length > 0 && !isOfferInReco) {
        history.push('/decouverte')
      }
    }
  }

  componentWillUnmount() {
    const { resetRecommendations } = this.props
    resetRecommendations()
  }

  handleFail = (state, action) => {
    this.setState(() => {
      throw new ApiError(action.payload.status)
    })
  }

  handleSuccess = (state, action) => {
    const { resetReadRecommendations, redirectToFirstRecommendationIfNeeded } = this.props

    const { data: loadedRecommendations = [] } = action && action.payload
    const atWorldsEnd = loadedRecommendations.length === 0

    this.setState({ atWorldsEnd, isLoading: false }, () => {
      resetReadRecommendations()
      redirectToFirstRecommendationIfNeeded(loadedRecommendations)
    })
  }

  updateRecommendations = () => {
    const {
      coordinates,
      currentRecommendation,
      loadRecommendations,
      readRecommendations,
      recommendations,
      shouldReloadRecommendations,
    } = this.props

    const { atWorldsEnd, isLoading } = this.state
    if (atWorldsEnd || isLoading) {
      return
    }

    this.setState({ isLoading: true }, () => {
      loadRecommendations(
        this.handleSuccess,
        this.handleFail,
        currentRecommendation,
        recommendations,
        readRecommendations,
        shouldReloadRecommendations,
        coordinates
      )
    })
  }

  render() {
    const { currentRecommendation, match } = this.props
    const { isLoading } = this.state

    return (
      <Fragment>
        {isLoading && <LoaderContainer />}

        {!isLoading && (
          <main className="discovery-page no-padding page">
            <Switch>
              <Route
                exact
                path={`${match.path}/:details(details|transition)?/:booking(reservation)?/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?/:confirmation(confirmation)?`}
                sensitive
              >
                <Route
                  exact
                  path={`${match.path}/:details(details|transition)?/:booking(reservation)?/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?`}
                  sensitive
                >
                  <DeckContainer handleRequestPutRecommendations={this.updateRecommendations} />
                </Route>
                <Route
                  exact
                  path={`${match.path}/:details(details)/:booking(reservation)/:bookingId([A-Z0-9]+)?`}
                  sensitive
                >
                  <BookingContainer recommendation={currentRecommendation} />
                </Route>
                <Route
                  exact
                  path={`${match.path}/:details(details)/:booking(reservation)/:bookingId([A-Z0-9]+)/:cancellation(annulation)/:confirmation(confirmation)`}
                  sensitive
                >
                  <BookingCancellationContainer />
                </Route>
              </Route>
              <Route>
                <PageNotFoundContainer />
              </Route>
            </Switch>
          </main>
        )}
      </Fragment>
    )
  }
}

Discovery.defaultProps = {
  currentRecommendation: null,
  readRecommendations: null,
  recommendations: null,
}

Discovery.propTypes = {
  coordinates: PropTypes.shape().isRequired,
  currentRecommendation: PropTypes.shape(),
  history: PropTypes.shape().isRequired,
  loadRecommendations: PropTypes.func.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      view: PropTypes.string,
      offerId: PropTypes.string,
    }),
    path: PropTypes.string.isRequired,
  }).isRequired,
  readRecommendations: PropTypes.arrayOf(PropTypes.shape()),
  recommendations: PropTypes.arrayOf(PropTypes.shape()),
  redirectToFirstRecommendationIfNeeded: PropTypes.func.isRequired,
  resetReadRecommendations: PropTypes.func.isRequired,
  resetRecommendations: PropTypes.func.isRequired,
  saveLastRecommendationsRequestTimestamp: PropTypes.func.isRequired,
  seedLastRequestTimestamp: PropTypes.number.isRequired,
  shouldReloadRecommendations: PropTypes.bool.isRequired,
  trackGeolocation: PropTypes.func.isRequired,
  updateLastRequestTimestamp: PropTypes.func.isRequired,
}

export default Discovery
