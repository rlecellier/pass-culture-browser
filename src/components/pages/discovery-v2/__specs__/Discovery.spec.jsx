import { shallow } from 'enzyme'
import React from 'react'
import { Route } from 'react-router'

import Discovery from '../Discovery'
import LoaderContainer from '../../../layout/Loader/LoaderContainer'
import AbsoluteFooterContainer from '../../../layout/AbsoluteFooter/AbsoluteFooterContainer'

describe('src | components | pages | discovery | Discovery', () => {
  let props

  beforeEach(() => {
    props = {
      currentRecommendation: {},
      deleteTutorials: jest.fn(),
      loadRecommendations: jest.fn(),
      location: {
        pathname: '',
        search: '',
      },
      match: {
        params: {},
      },
      readRecommendations: [],
      recommendations: [],
      redirectToFirstRecommendationIfNeeded: jest.fn(),
      resetReadRecommendations: jest.fn(),
      resetRecommendations: jest.fn(),
      resetSeenOffers: jest.fn(),
      resetRecommendationsAndBookings: jest.fn(),
      saveLastRecommendationsRequestTimestamp: jest.fn(),
      seedLastRequestTimestamp: 1574236357670,
      seenOffers: [],
      shouldReloadRecommendations: false,
      tutorials: [],
      updateLastRequestTimestamp: jest.fn(),
    }
  })

  it('should match the snapshot', () => {
    // given
    const wrapper = shallow(<Discovery {...props} />)

    // then
    expect(wrapper).toMatchSnapshot()
  })

  describe('constructor', () => {
    it('should initialize state with default values', () => {
      // given
      const wrapper = shallow(<Discovery {...props} />)

      // then
      expect(wrapper.state()).toStrictEqual({
        atWorldsEnd: false,
        hasError: false,
        hasError500: false,
        hasNoMoreRecommendations: false,
        isLoading: false,
      })
    })
  })

  describe('when mount', () => {
    it('should load recommendations when it is required to fetch recommendations', () => {
      // given
      props.shouldReloadRecommendations = true

      // when
      shallow(<Discovery {...props} />)

      // then
      expect(props.loadRecommendations).toHaveBeenCalledWith(
        expect.any(Function),
        expect.any(Function),
        props.currentRecommendation,
        props.recommendations,
        props.readRecommendations,
        props.seenOffers,
        props.shouldReloadRecommendations
      )
      expect(props.saveLastRecommendationsRequestTimestamp).toHaveBeenCalledWith()
    })

    it('should redirect to first recommendation when it is not required to fetch recommendations', () => {
      // given
      props.shouldReloadRecommendations = false

      // when
      shallow(<Discovery {...props} />)

      // then
      expect(props.loadRecommendations).not.toHaveBeenCalledWith()
      expect(props.saveLastRecommendationsRequestTimestamp).not.toHaveBeenCalledWith()
      expect(props.redirectToFirstRecommendationIfNeeded).toHaveBeenCalledWith(
        props.recommendations
      )
    })

    it('should display discovery when API is up and data are filled', () => {
      // given
      const wrapper = shallow(<Discovery {...props} />)

      // when
      wrapper.setState({
        hasError: false,
        hasNoMoreRecommendations: false,
        isLoading: false,
      })

      // then
      expect(wrapper.find(Route)).toHaveLength(2)
      expect(wrapper.find(AbsoluteFooterContainer)).toHaveLength(1)
      expect(wrapper.find(LoaderContainer)).toHaveLength(1)
    })

    it('should display error message when API is down', () => {
      // given
      const wrapper = shallow(<Discovery {...props} />)

      // when
      wrapper.setState({
        hasError: true,
        hasNoMoreRecommendations: true,
        isLoading: false,
      })

      // then
      expect(wrapper.find(Route)).toHaveLength(0)
      expect(wrapper.find(AbsoluteFooterContainer)).toHaveLength(0)
      expect(wrapper.find(LoaderContainer)).toHaveLength(1)
    })

    it('should display loading message when API is slow', () => {
      // given
      const wrapper = shallow(<Discovery {...props} />)

      // when
      wrapper.setState({
        hasError: false,
        hasNoMoreRecommendations: true,
        isLoading: true,
      })

      // then
      expect(wrapper.find(Route)).toHaveLength(0)
      expect(wrapper.find(AbsoluteFooterContainer)).toHaveLength(0)
      expect(wrapper.find(LoaderContainer)).toHaveLength(1)
    })

    it('should display no offer message when API returns zero offer', () => {
      // given
      const wrapper = shallow(<Discovery {...props} />)

      // when
      wrapper.setState({
        hasError: false,
        hasNoMoreRecommendations: true,
        isLoading: false,
      })

      // then
      expect(wrapper.find(Route)).toHaveLength(0)
      expect(wrapper.find(AbsoluteFooterContainer)).toHaveLength(0)
      expect(wrapper.find(LoaderContainer)).toHaveLength(1)
    })
  })

  describe('when unmount', () => {
    it('should delete tutorials for current user', () => {
      // given
      props.tutorials = [
        {
          id: 'hello',
          productOrTutoIdentifier: 'tuto_0',
        },
      ]
      const wrapper = shallow(<Discovery {...props} />)

      // when
      wrapper.unmount()

      // then
      expect(props.deleteTutorials).toHaveBeenCalledWith([
        {
          id: 'hello',
          productOrTutoIdentifier: 'tuto_0',
        },
      ])
    })

    it('should not delete tutorials for current user', () => {
      // given
      props.tutorials = []
      const wrapper = shallow(<Discovery {...props} />)

      // when
      wrapper.unmount()

      // then
      expect(props.deleteTutorials).not.toHaveBeenCalledWith()
    })
  })

  describe('functions', () => {
    describe('handleSuccess', () => {
      it('should reset read recommendations', () => {
        // given
        const action = {
          payload: {
            data: [{ id: 'ABC1' }],
          },
        }
        const wrapper = shallow(<Discovery {...props} />)

        // when
        wrapper.instance().handleSuccess({}, action)

        // then
        expect(props.resetReadRecommendations).toHaveBeenCalledWith()
      })

      it('should reset seen offers', () => {
        // given
        const action = {
          payload: {
            data: [{ id: 'AB' }],
          },
        }
        const wrapper = shallow(<Discovery {...props} />)

        // when
        wrapper.instance().handleSuccess({}, action)

        // then
        expect(props.resetSeenOffers).toHaveBeenCalledWith()
      })

      it('should set atWorldsEnd state to true when no recommendations have been fetched', () => {
        // given
        const action = {
          payload: {
            data: [],
          },
        }
        props.recommendations = [{ id: 'ABC1' }]
        const wrapper = shallow(<Discovery {...props} />)

        // when
        wrapper.instance().handleSuccess({}, action)

        // then
        expect(wrapper.state()).toStrictEqual({
          atWorldsEnd: true,
          hasError: false,
          hasError500: false,
          hasNoMoreRecommendations: false,
          isLoading: false,
        })
      })

      it('should set atWorldsEnd state to false when recommendations have been fetched', () => {
        // given
        const action = {
          payload: {
            data: [{ id: 'DEF2' }],
          },
        }
        props.recommendations = [{ id: 'ABC1' }]
        const wrapper = shallow(<Discovery {...props} />)

        // when
        wrapper.instance().handleSuccess({}, action)

        // then
        expect(wrapper.state()).toStrictEqual({
          atWorldsEnd: false,
          hasError: false,
          hasError500: false,
          hasNoMoreRecommendations: false,
          isLoading: false,
        })
      })

      it('should set hasNoMoreRecommendations state to true when no recommendations have been fetched and no recommendations are loaded', () => {
        // given
        const action = {
          payload: {
            data: [],
          },
        }
        props.recommendations = []
        const wrapper = shallow(<Discovery {...props} />)

        // when
        wrapper.instance().handleSuccess({}, action)

        // then
        expect(wrapper.state()).toStrictEqual({
          atWorldsEnd: true,
          hasError: false,
          hasError500: false,
          hasNoMoreRecommendations: true,
          isLoading: false,
        })
      })
    })
  })

  describe('when update', () => {
    it('should update seed last request timestamp when date is posterior to limit', () => {
      // given
      jest
        .spyOn(global.Date, 'now')
        .mockImplementationOnce(() => new Date('2019-11-20T13:00:00.000Z').valueOf())

      // when
      const wrapper = shallow(<Discovery {...props} />)
      props.location = {
        pathname: 'nouveauPath',
        search: '',
      }
      wrapper.setProps(props)

      // then
      expect(props.updateLastRequestTimestamp).toHaveBeenCalledWith()
    })

    it('should not update seed last request timestamp when date is anterior to limit', () => {
      // given
      jest
        .spyOn(global.Date, 'now')
        .mockImplementationOnce(() => new Date('2019-11-20T00:00:00.000Z').valueOf())

      // when
      const wrapper = shallow(<Discovery {...props} />)
      props.location = {
        pathname: 'nouveauPath',
        search: '',
      }
      wrapper.setProps(props)

      // then
      expect(props.updateLastRequestTimestamp).not.toHaveBeenCalledWith()
    })
  })
})
