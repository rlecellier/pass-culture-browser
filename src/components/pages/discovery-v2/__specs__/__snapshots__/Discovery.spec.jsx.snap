// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | discovery | Discovery should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Discovery
    currentRecommendation={Object {}}
    deleteTutorials={[MockFunction]}
    loadRecommendations={[MockFunction]}
    location={
      Object {
        "pathname": "",
        "search": "",
      }
    }
    match={
      Object {
        "params": Object {},
      }
    }
    readRecommendations={Array []}
    recommendations={Array []}
    redirectToFirstRecommendationIfNeeded={
      [MockFunction] {
        "calls": Array [
          Array [
            Array [],
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    resetReadRecommendations={[MockFunction]}
    resetRecommendations={[MockFunction]}
    resetRecommendationsAndBookings={[MockFunction]}
    resetSeenOffers={[MockFunction]}
    saveLastRecommendationsRequestTimestamp={[MockFunction]}
    seedLastRequestTimestamp={1574236357670}
    seenOffers={Array []}
    shouldReloadRecommendations={false}
    tutorials={Array []}
    updateLastRequestTimestamp={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <main
          className="discovery-page no-padding page with-footer"
        >
          <Route
            path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details|transition)?/:booking(reservation)?/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?"
            render={[Function]}
            sensitive={true}
          />
          <Route
            path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details)/:booking(reservation)/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?/:confirmation(confirmation)?"
            render={[Function]}
            sensitive={true}
          />
          <withRouter(AbsoluteFooter)
            areDetailsVisible={false}
            borderTop={true}
            id="deck-footer"
          />
        </main>,
        <withRouter(Loader)
          hasError={false}
          hasError500={false}
          hasNoMoreRecommendations={false}
          isLoading={false}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Route
              path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details|transition)?/:booking(reservation)?/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?"
              render={[Function]}
              sensitive={true}
            />,
            <Route
              path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details)/:booking(reservation)/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?/:confirmation(confirmation)?"
              render={[Function]}
              sensitive={true}
            />,
            <withRouter(AbsoluteFooter)
              areDetailsVisible={false}
              borderTop={true}
              id="deck-footer"
            />,
          ],
          "className": "discovery-page no-padding page with-footer",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "route-discovery-deck",
            "nodeType": "class",
            "props": Object {
              "path": "/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details|transition)?/:booking(reservation)?/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?",
              "render": [Function],
              "sensitive": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "route-discovery-booking",
            "nodeType": "class",
            "props": Object {
              "path": "/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details)/:booking(reservation)/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?/:confirmation(confirmation)?",
              "render": [Function],
              "sensitive": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "areDetailsVisible": false,
              "borderTop": true,
              "id": "deck-footer",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "main",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "hasError": false,
          "hasError500": false,
          "hasNoMoreRecommendations": false,
          "isLoading": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <main
            className="discovery-page no-padding page with-footer"
          >
            <Route
              path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details|transition)?/:booking(reservation)?/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?"
              render={[Function]}
              sensitive={true}
            />
            <Route
              path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details)/:booking(reservation)/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?/:confirmation(confirmation)?"
              render={[Function]}
              sensitive={true}
            />
            <withRouter(AbsoluteFooter)
              areDetailsVisible={false}
              borderTop={true}
              id="deck-footer"
            />
          </main>,
          <withRouter(Loader)
            hasError={false}
            hasError500={false}
            hasNoMoreRecommendations={false}
            isLoading={false}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Route
                path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details|transition)?/:booking(reservation)?/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?"
                render={[Function]}
                sensitive={true}
              />,
              <Route
                path="/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details)/:booking(reservation)/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?/:confirmation(confirmation)?"
                render={[Function]}
                sensitive={true}
              />,
              <withRouter(AbsoluteFooter)
                areDetailsVisible={false}
                borderTop={true}
                id="deck-footer"
              />,
            ],
            "className": "discovery-page no-padding page with-footer",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "route-discovery-deck",
              "nodeType": "class",
              "props": Object {
                "path": "/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details|transition)?/:booking(reservation)?/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?",
                "render": [Function],
                "sensitive": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "route-discovery-booking",
              "nodeType": "class",
              "props": Object {
                "path": "/decouverte/:offerId(tuto|[A-Z0-9]+)/:mediationId(vide|fin|[A-Z0-9]+)/:details(details)/:booking(reservation)/:bookingId([A-Z0-9]+)?/:cancellation(annulation)?/:confirmation(confirmation)?",
                "render": [Function],
                "sensitive": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "areDetailsVisible": false,
                "borderTop": true,
                "id": "deck-footer",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "main",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "hasError": false,
            "hasError500": false,
            "hasNoMoreRecommendations": false,
            "isLoading": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
